<br>

<img src="./image/Logo/logo_elia_group.png" width = 200>

<br>

# Conditional Selections: Filtering
<br>

One of the most beloved and useful function in excel is the filter function - and of course, you can do the same in Python. You can use conditional selections to select specific rows and narrow your analysis down. And to make things easier, you can save selections you plan to use often as their own variables. Let's get right to it!

<img src= "./image/conditional_selections.png" width = 300>

First, let's have a look at the dataset "physical flow" again: 

In [1]:
import pandas as pd

In [2]:
energy_flow = pd.read_csv("./data/energy/physical_flow_2021_1_01.csv", sep = ";", parse_dates = True, index_col = 0)
energy_flow.head()

Resolution code   Control area  Physical Flow Value
Datetime                                                                     
2021-12-01 23:45:00+01:00           PT15M    Netherlands             -419.704
2021-12-01 23:45:00+01:00           PT15M  UnitedKingdom             1021.774
2021-12-01 23:45:00+01:00           PT15M     Luxembourg              -53.708
2021-12-01 23:45:00+01:00           PT15M        Germany            -1001.788
2021-12-01 23:45:00+01:00           PT15M         France              581.908

As you can see, the data set describes different physical flow values from different countries (the neighbouring bidding zones of Belgium), measured at similar datetimes.
Let's check which countries are represented here:

In [3]:
energy_flow["Control area"].unique()

array(['Netherlands', 'UnitedKingdom', 'Luxembourg', 'Germany', 'France'],
      dtype=object)

Imagine you want to take a closer look at France and Luxembourg. To do so, you need to select the column of interest and "filter" your area of choice with a conditional statement which returns either True or False. This is then used to filter your data set. 

In [4]:
france = energy_flow[energy_flow["Control area"] == "France"]

In [5]:
luxembourg = energy_flow[energy_flow["Control area"] == "Luxembourg"]

In [6]:
france.head(n=3)

Resolution code Control area  Physical Flow Value
Datetime                                                                   
2021-12-01 23:45:00+01:00           PT15M       France              581.908
2021-12-01 23:30:00+01:00           PT15M       France              691.116
2021-12-01 23:15:00+01:00           PT15M       France              523.224

In [7]:
luxembourg.head(n=3)

Resolution code Control area  Physical Flow Value
Datetime                                                                   
2021-12-01 23:45:00+01:00           PT15M   Luxembourg              -53.708
2021-12-01 23:30:00+01:00           PT15M   Luxembourg              -57.983
2021-12-01 23:15:00+01:00           PT15M   Luxembourg              -24.082

### Excercise

1. Look at the dataframe `energy_flow` above and select the control area "Germany". 
2. Save your selection into a variable called `germany`.
3. Look at the first three rows to see if it worked. 

Since the data set includes hourly data from one day, you could calculate the mean physical flow of that day per selected country:

In [8]:
print('Mean Physical Flow of France in MW: ', round(france['Physical Flow Value'].mean()))

Mean Physical Flow of France in MW:  418


In [9]:
print('Mean Physical Flow of Luxembourg in MW: ', round(luxembourg['Physical Flow Value'].mean()))

Mean Physical Flow of Luxembourg in MW:  69


Since a positive figure means export from Belgium, you can now use these to calculate more targeted metrics: 

In [10]:
print('On that day, Belgium exports {} MW on average more to France compared to Luxembourg.'\
      .format(round(france['Physical Flow Value'].mean() - luxembourg['Physical Flow Value'].mean(),2)))

On that day, Belgium exports 348.28 MW on average more to France compared to Luxembourg.


## Advanced Conditionals: Using Masks 
<br>

Sure, it is nice to filter just one thing. But what if you want **to filter on > 1 criteria**? Then it can be easier to use a mask. No, not a face mask. Rather a boolean mask. <br>

Imagine you would like to not just select France OR Luxembourg, but both countries as well as Germany. With a mask, you specify these multiple conditions. Your mask evaluates the different conditions and returns either TRUE/FALSE. In a second step this mask is used as a filter as you have already learned in Filtering.
The pipe operator `|` is used as on OR whereas the `&` is used as an AND. But enough talking, let's try it out! 

1. Create a variable that stores all the conditions you would like to choose

In [11]:
countries_mask = (energy_flow["Control area"] == "France") | (energy_flow["Control area"] == "Luxembourg") | (energy_flow["Control area"] == "Germany")

2. Look at your mask. It returns whether your conditions have been met for each row or not

In [12]:
countries_mask

Datetime
2021-12-01 23:45:00+01:00    False
2021-12-01 23:45:00+01:00    False
2021-12-01 23:45:00+01:00     True
2021-12-01 23:45:00+01:00     True
2021-12-01 23:45:00+01:00     True
                             ...  
2021-12-01 00:00:00+01:00     True
2021-12-01 00:00:00+01:00     True
2021-12-01 00:00:00+01:00    False
2021-12-01 00:00:00+01:00    False
2021-12-01 00:00:00+01:00     True
Name: Control area, Length: 480, dtype: bool

3. Now you can directly access your mask, using: 

In [13]:
selected_countries = energy_flow[countries_mask]

In [14]:
selected_countries["Control area"].unique()

array(['Luxembourg', 'Germany', 'France'], dtype=object)

&#128526; nice, well done!

Let's create another mask just for fun. Now we want to have all the data related to a physical flow is higher 1000 MW and find out whether it is import or export. For that, let's have a look at the original dataframe again: 

In [15]:
energy_flow.head()

Resolution code   Control area  Physical Flow Value
Datetime                                                                     
2021-12-01 23:45:00+01:00           PT15M    Netherlands             -419.704
2021-12-01 23:45:00+01:00           PT15M  UnitedKingdom             1021.774
2021-12-01 23:45:00+01:00           PT15M     Luxembourg              -53.708
2021-12-01 23:45:00+01:00           PT15M        Germany            -1001.788
2021-12-01 23:45:00+01:00           PT15M         France              581.908

Let's define that in this case, a high physical flow means < - 1000 MW and > 1000 MW.

In [16]:
high_flow_mask = (energy_flow["Physical Flow Value"] < -1000) | (energy_flow["Physical Flow Value"] > 1000)

In [17]:
high_flow_mask

Datetime
2021-12-01 23:45:00+01:00    False
2021-12-01 23:45:00+01:00     True
2021-12-01 23:45:00+01:00    False
2021-12-01 23:45:00+01:00     True
2021-12-01 23:45:00+01:00    False
                             ...  
2021-12-01 00:00:00+01:00    False
2021-12-01 00:00:00+01:00    False
2021-12-01 00:00:00+01:00    False
2021-12-01 00:00:00+01:00     True
2021-12-01 00:00:00+01:00    False
Name: Physical Flow Value, Length: 480, dtype: bool

In [18]:
high_flow = energy_flow[high_flow_mask]

In [19]:
high_flow.head()

Resolution code   Control area  Physical Flow Value
Datetime                                                                     
2021-12-01 23:45:00+01:00           PT15M  UnitedKingdom             1021.774
2021-12-01 23:45:00+01:00           PT15M        Germany            -1001.788
2021-12-01 23:30:00+01:00           PT15M  UnitedKingdom             1023.455
2021-12-01 23:30:00+01:00           PT15M        Germany            -1002.104
2021-12-01 23:15:00+01:00           PT15M  UnitedKingdom             1023.362

## Groupby
<br> 

One of the most flexible ways to group your data and aggregate in pandas is with `.groupby()`. So what does this actually mean? Let's have a look at the following example:

In [20]:
energy_flow.groupby("Control area").mean()

Physical Flow Value
Control area                      
France                  417.687250
Germany                -692.056208
Luxembourg               69.402396
Netherlands             186.888375
UnitedKingdom           833.411906

As you can see from the example above, `groupby()` groups your data by the column(s) that you hand over to the function. In this case "Control area". In addition, `groupby()` **only works with an aggregator** such as sum() or mean(). This means, you have to tell the function what to do with each group. In this case, calculate the mean using `mean()`. Also notice, that the **column you grouped on/by becomes your new index**!

**Question:**
Do you know why only the column "Physical Flow Value" is displayed in our example?

You can do many more cool things. If you want to change the order in which the aggregated values are displayed, you can just chain the command `.sort_values` to your groupby statement. In general, you can use `.sort_values` for sorting any  column of a DataFrame.

In [21]:
energy_flow.groupby("Control area").mean().sort_values(by=["Physical Flow Value"], ascending=False)

Physical Flow Value
Control area                      
UnitedKingdom           833.411906
France                  417.687250
Netherlands             186.888375
Luxembourg               69.402396
Germany                -692.056208

For our next example, let's have a look at a bigger and more complex data set. To do so, you first have to import the csv: 

In [22]:
pf_high_voltage = pd.read_csv("./data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";", index_col = 0)

In [23]:
pf_high_voltage.head()

Resolution code Asset Elia ID  \
Datetime                                                  
2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2022-05-30T23:15:00+02:00           PT15M        380_29   
2022-05-30T23:15:00+02:00           PT15M       380_111   
2022-05-30T23:15:00+02:00           PT15M        380_79   

                                         Asset name     Asset type  \
Datetime                                                             
2022-05-30T23:15:00+02:00      DOEL-MERCATOR 380.51  Internal Line   
2022-05-30T23:15:00+02:00     HORTA-MERCATOR 380.74  Internal Line   
2022-05-30T23:15:00+02:00  RILLAND-ZANDVLIET 380.29        Tieline   
2022-05-30T23:15:00+02:00    GEZELLE-STEVIN 380.111  Internal Line   
2022-05-30T23:15:00+02:00   AVELGEM-MASTAING 380.79        Tieline   

                           Base Voltage AC/DC  Physical Flow  Loading  \
Datetime                                                                
2022-05-30T23:15:00+02:00         380.0    AC      13.754076     0.68   
2022-05-30T23:15:00+02:00         380.0    AC     238.395599     7.01   
2022-05-30T23:15:00+02:00         380.0    AC     232.342749    10.78   
2022-05-30T23:15:00+02:00         380.0    AC     121.828900     3.72   
2022-05-30T23:15:00+02:00         380.0    AC       0.015625     0.00   

                                                   Asset geographic location  \
Datetime                                                                       
2022-05-30T23:15:00+02:00  {"coordinates": [[4.28, 51.23], [4.3072839, 51...   
2022-05-30T23:15:00+02:00  {"coordinates": [[3.6036937, 51.11457], [3.9, ...   
2022-05-30T23:15:00+02:00  {"coordinates": [[4.181, 51.416], [4.24811, 51...   
2022-05-30T23:15:00+02:00  {"coordinates": [[3.2011967, 51.2638468], [3.1...   
2022-05-30T23:15:00+02:00  {"coordinates": [[3.4690219, 50.7824407], [3.4...   

                                                    geo_point_2d  
Datetime                                                          
2022-05-30T23:15:00+02:00                  51.2655739,4.29364195  
2022-05-30T23:15:00+02:00  51.138284999999996,3.7518468499999997  
2022-05-30T23:15:00+02:00                   51.39276245,4.214555  
2022-05-30T23:15:00+02:00                    51.29485,3.19246695  
2022-05-30T23:15:00+02:00                 50.59037085,3.43451095

The data set above describes the physical flow on the Belgian 380-kV lines (high-voltage) and on the interconnections with the neighboring countries. The "Loading" indicates how heavily the line is loaded relative to the maximum possible line loading. For the purpose of this training, you look at the data from only one day - the 30th of May 2022. The "Physical Flow" is given in MW whereas the "Loading" is given in %. 

Let's use what we have learned so far: 

In [24]:
pf_high_voltage.groupby("Asset name").mean().head()

Base Voltage  Physical Flow    Loading
Asset name                                                   
ACHENE-GRAMME 380.10          380.0     342.429992  21.129011
ACHENE-LONNY 380.19           380.0     295.404589  18.932418
AUBANGE-BRUME 380.13          380.0     117.085284  13.561868
AVELGEM-HORTA 380.101         380.0     198.139137   6.929011
AVELGEM-HORTA 380.102         380.0     201.550017   7.048901

**Tip:** Sometimes, if you want to aggregate different columns in different ways and to make your code cleaner, it is best to move the aggregations out and store them as a dictionary. 

In [25]:
aggs = {
    'Physical Flow': 'mean',
    'Loading': 'max'
}

pf_high_voltage.groupby('Asset name').agg(aggs).sort_values(by=['Physical Flow','Loading']).tail()

Physical Flow  Loading
Asset name                                        
MAASBRACHT-VAN EYCK 380.28     498.400809    58.51
DOEL-MERCATOR 380.53           519.729299    45.17
DOEL-ZANDVLIET 380.26          548.510699    74.14
DOEL-MERCATOR 380.54           642.709944    51.01
LILLO-ZANDVLIET 380.66        3159.260010   202.13

If you need to aggregate a certain column in several different ways, you can store the column and the different aggregators as key-value pairs in a dictionary.

In [26]:
aggs2 = {
    'Loading': ['min', 'mean', 'max', 'std']
}

loading_stats = pf_high_voltage.groupby('Asset name').agg(aggs2)

In [27]:
loading_stats.head()

Loading                             
                          min       mean    max        std
Asset name                                                
ACHENE-GRAMME 380.10     0.61  21.129011  47.73  14.050990
ACHENE-LONNY 380.19      0.43  18.932418  45.76  14.399867
AUBANGE-BRUME 380.13     8.33  13.561868  18.27   2.392377
AVELGEM-HORTA 380.101    0.35   6.929011  16.23   4.541299
AVELGEM-HORTA 380.102    0.36   7.048901  16.27   4.516142

As you can see, this multi-aggregation creates a **multi-index**. Multi-indexes can be difficult to work with. But no worries, there is an easy way to deal with it. For instance, you can **drop the top level**. In this case "Loading": 

In [28]:
loading_stats.columns = loading_stats.columns.droplevel(level = 0)

In [29]:
loading_stats.head()

min       mean    max        std
Asset name                                              
ACHENE-GRAMME 380.10   0.61  21.129011  47.73  14.050990
ACHENE-LONNY 380.19    0.43  18.932418  45.76  14.399867
AUBANGE-BRUME 380.13   8.33  13.561868  18.27   2.392377
AVELGEM-HORTA 380.101  0.35   6.929011  16.23   4.541299
AVELGEM-HORTA 380.102  0.36   7.048901  16.27   4.516142

It might be necessary to rename the new "columns", so you keep in mind that they are all stats of "loading".

In [30]:
loading_stats.columns = ["loading_min", "loading_mean", "loading_max", "loading_std"]

In [31]:
loading_stats.head()

loading_min  loading_mean  loading_max  loading_std
Asset name                                                                
ACHENE-GRAMME 380.10          0.61     21.129011        47.73    14.050990
ACHENE-LONNY 380.19           0.43     18.932418        45.76    14.399867
AUBANGE-BRUME 380.13          8.33     13.561868        18.27     2.392377
AVELGEM-HORTA 380.101         0.35      6.929011        16.23     4.541299
AVELGEM-HORTA 380.102         0.36      7.048901        16.27     4.516142

And now, you can sort by any of the columns. Here, by average loading:

In [32]:
loading_stats.sort_values(by='loading_mean', ascending=False).head()

loading_min  loading_mean  loading_max  loading_std
Asset name                                                                 
LILLO-ZANDVLIET 380.66       190.20    195.449200       202.13     5.982019
DOEL-MERCATOR 380.54          28.50     40.869231        51.01     5.367424
GEZELLE-NEMO 380.114           2.00     37.317294        76.68    22.112781
DOEL-ZANDVLIET 380.26         13.67     34.723846        74.14    16.444268
LIXHE-ALEGRO 380.115           2.66     33.206667        96.09    25.622747

## Selecting the max and min values with Index Max and Min
<br>

The last cool thing that is definitely worth learning in the beginning is `idxmin` and `idxmax`. In addition to `.max()` and `.min()`, which returns the maximum or minimum values, you can use `.idxmax()` and `.idxmin()` to return the *index* pertaining to the maximum and minimum values. <br>

For example, let's use `.idxmax()` to find the "Asset name" with the highest standard deviation in its loading:

In [33]:
loading_stats["loading_std"].idxmax()

'LIXHE-ALEGRO 380.115'

<br>

## Recap, Tips & Takeaways &#128161;

<br>

<div class="alert alert-block alert-success">

**Let's see what might be cool to keep in mind:**

- there are two ways to combine dataframes: `pd.merge()` and `pd.concat()`
- you can get quick stats with `df_name.describe()`
- `df_name["column_name"].unique()` lists all the unique values within a column 
- filters in Python are a boolean conditional selection: `df_name[df_name["column_name"] == "target_value"]`
- multiple filters can be linked together with `|` and  `&` statement
- you can define aggregators for several columns with a dictionary: <br>
    
    aggs = {
        'column_1': 'mean',
        'column_2': 'max'
    }
        
</div>